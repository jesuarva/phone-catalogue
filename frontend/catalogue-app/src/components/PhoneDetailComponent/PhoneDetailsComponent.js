import PropTypes from 'prop-types';
import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Container, Jumbotron, Table } from 'reactstrap';
import Pic from '../Pic/Pic';

export const PhoneDetailsComponent = (props) => {
  const { data } = props;

  // Get url-param and repalces dach-character with 'spaces'
  const model = props.match.params.model.replace(/-/g, ' ');

  // Get specific phone-details
  const phone = data.filter((phone) => phone.Model === model)[0];
  const { Model, image } = phone;

  let details = [];

  // Buil the details to display
  for (let key in phone) {
    if (key === 'image' || key === 'Model') continue;
    details.push(
      <tr key={Math.random() + Date.now()}>
        <th scope="row">{key}</th>
        <td>{phone[key]}</td>
      </tr>,
    );
  }

  return (
    <Fragment>
      <div className="row">
        <div className="col-12">
          <Link to="/">
            <h1>Back to catalogue</h1>
          </Link>
        </div>
      </div>
      <div className="row">
        <div className="col-12 col-4 card-phone-image">
          <Pic img={image} altAttribute={Model} />
        </div>
        <div className="col-12 col-8 card-phone-details">
          <Jumbotron fluid>
            <Container fluid>
              <h1 className="display-3">{Model}</h1>
              <Table>
                <tbody>{details}</tbody>
              </Table>
            </Container>
          </Jumbotron>
        </div>
      </div>
    </Fragment>
  );
};

PhoneDetailsComponent.propTypes = {
  data: PropTypes.arrayOf(
    PropTypes.shape({
      Model: PropTypes.string.isRequired,
      image: PropTypes.string.isRequired,
    }),
  ).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      model: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
};

const mapStateToProps = (state) => {
  return {
    data: state.data,
  };
};

export default connect(mapStateToProps)(PhoneDetailsComponent);
