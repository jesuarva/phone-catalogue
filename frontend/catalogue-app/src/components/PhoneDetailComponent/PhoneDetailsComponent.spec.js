import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router, Link } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { Container, Jumbotron, Table } from 'reactstrap';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import mainReducer from '../../reducers/index';
import TestWrapper from '../TestWrapper/TestWrapper';
import { PhoneDetailsComponent } from './PhoneDetailsComponent';

configure({ adapter: new Adapter() });

describe('PhoneDetailsComponent component', () => {
  function setUp() {
    const props = {
      data: [
        {
          image:
            'https://store.storeimages.cdn-apple.com/4981/as-images.apple.com/is/image/AppleInc/aos/published/images/i/ph/iphone8/select/iphone8-select-2017?wid=212&hei=292&fmt=png-alpha&.v=1503618523038',
          Model: 'Apple iPhone 8 Plus',
        },
      ],
      match: {
        params: {
          model: 'Apple-iPhone-8-Plus',
        },
      },
    };

    const store = createStore(mainReducer, applyMiddleware(thunk));
    const component = (
      <Provider store={store}>
        <Router>
          <PhoneDetailsComponent {...props} />
        </Router>
      </Provider>
    );

    return {
      props,
      component,
    };
  }

  it('should renders correctly', () => {
    const { component } = setUp();
    const tree = renderer.create(component);
    expect(tree).toMatchSnapshot();
  });

  describe('Renders children component', () => {
    const { component } = setUp();
    const mounted = mount(<TestWrapper>{() => component}</TestWrapper>);

    it('should renders a Link component', () => {
      const link = mounted.find(Link);
      expect(link).toHaveLength(1);
    });

    it('should renders a Jumbotron component', () => {
      const jumbo = mounted.find(Jumbotron);
      expect(jumbo).toHaveLength(1);
    });

    it('should renders a Container component', () => {
      const container = mounted.find(Container);
      expect(container).toHaveLength(1);
    });

    it('should renders a Table component', () => {
      const table = mounted.find(Table);
      expect(table).toHaveLength(1);
    });
  });
});
