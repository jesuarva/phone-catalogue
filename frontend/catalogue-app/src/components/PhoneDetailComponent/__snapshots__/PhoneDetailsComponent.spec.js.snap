// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoneDetailsComponent component should renders correctly 1`] = `
Array [
  <div
    className="row"
  >
    <div
      className="col-12"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <h1>
          Back to catalogue
        </h1>
      </a>
    </div>
  </div>,
  <div
    className="row"
  >
    <div
      className="col-12 col-4 card-phone-image"
    >
      <div
        className="pic"
      >
        <figure
          className="figure"
        >
          <img
            alt="Apple iPhone 8 Plus"
            className="figure-img img-fluid rounded"
            src="https://store.storeimages.cdn-apple.com/4981/as-images.apple.com/is/image/AppleInc/aos/published/images/i/ph/iphone8/select/iphone8-select-2017?wid=212&hei=292&fmt=png-alpha&.v=1503618523038"
          />
        </figure>
      </div>
    </div>
    <div
      className="col-12 col-8 card-phone-details"
    >
      <div
        className="jumbotron jumbotron-fluid"
      >
        <div
          className="container-fluid"
        >
          <h1
            className="display-3"
          >
            Apple iPhone 8 Plus
          </h1>
          <table
            className="table"
          >
            <tbody />
          </table>
        </div>
      </div>
    </div>
  </div>,
]
`;
